File: cubesat_gs/web/lttb.py
"""Largest-Triangle-Three-Buckets downsampling (Steinarsson 2013) for time series."""
from __future__ import annotations


def lttb(points: list[tuple[float, float]], threshold: int) -> list[tuple[float, float]]:
    n = len(points)
    if threshold >= n or threshold < 3 or n < 3:
        return list(points)
    sampled = [points[0]]
    bucket = (n - 2) / (threshold - 2)
    a = 0
    for i in range(threshold - 2):
        r0 = int((i + 1) * bucket) + 1
        r1 = min(int((i + 2) * bucket) + 1, n)
        avg_x = sum(p[0] for p in points[r0:r1]) / max(r1 - r0, 1)
        avg_y = sum(p[1] for p in points[r0:r1]) / max(r1 - r0, 1)
        s0 = int(i * bucket) + 1
        s1 = int((i + 1) * bucket) + 1
        ax, ay = points[a]
        best, best_area = s0, -1.0
        for j in range(s0, s1):
            area = abs((ax - avg_x) * (points[j][1] - ay) - (ax - points[j][0]) * (avg_y - ay)) * 0.5
            if area > best_area:
                best, best_area = j, area
        sampled.append(points[best])
        a = best
    sampled.append(points[-1])
    return sampled

File: cubesat_gs/web/frontend/src/lib/lttb.ts
/** Largest-Triangle-Three-Buckets downsampling (Steinarsson 2013) for time series. Mirrors cubesat_gs/web/lttb.py. */
export function lttb(points: [number, number][], threshold: number): [number, number][] {
  const n = points.length;
  if (threshold >= n || threshold < 3 || n < 3) return points.slice();

  const sampled: [number, number][] = [points[0]];
  const bucket = (n - 2) / (threshold - 2);
  let a = 0;
  for (let i = 0; i < threshold - 2; i++) {
    const r0 = Math.floor((i + 1) * bucket) + 1;
    const r1 = Math.min(Math.floor((i + 2) * bucket) + 1, n);
    let sumX = 0, sumY = 0;
    for (let k = r0; k < r1; k++) { sumX += points[k][0]; sumY += points[k][1]; }
    const count = Math.max(r1 - r0, 1);
    const avgX = sumX / count, avgY = sumY / count;

    const s0 = Math.floor(i * bucket) + 1;
    const s1 = Math.floor((i + 1) * bucket) + 1;
    const [ax, ay] = points[a];
    let best = s0, bestArea = -1;
    for (let j = s0; j < s1; j++) {
      const area = Math.abs((ax - avgX) * (points[j][1] - ay) - (ax - points[j][0]) * (avgY - ay)) * 0.5;
      if (area > bestArea) { best = j; bestArea = area; }
    }
    sampled.push(points[best]);
    a = best;
  }
  sampled.push(points[n - 1]);
  return sampled;
}
